### DISCUSSION TITLE
Is it possible to assign a raw disk?

### DISCUSSION THREAD

**Discussion by windowsrefund** [vagrant-libvirt/vagrant-libvirt#1261] [Q&A]
Hello,

I've read through the documentation but am still not clear on whether it will be possible to assign a raw disk. Before considering Vagrant, I was able to achieve this by editing my node's XML (where <UNIQUE_ID> represents a unique disk on my system...

```
<disk type='block' device='disk'>
      <driver name='qemu' type='raw'/>
      <source dev='/dev/disk/by-id/<UNIQUE_ID>'/>
      <target dev='vdb' bus='virtio'/>
...
```

Basically, the node sees and can use /dev/vdb which is the unique raw disk. Thanks in advance for any assistance.

**Comment by electrofelix** [comment]
@windowsrefund not as it currently stands. Basically vagrant-libvirt is taking the additional disks and iterating over the using the template at https://github.com/vagrant-libvirt/vagrant-libvirt/blob/master/lib/vagrant-libvirt/templates/domain.xml.erb#L128-L152.

No reason that couldn't be updated to support additional options, would need some adjustment to ensure defaults applied for the current standard approach of adding disks, or if support for a callback was added for amending the domain XML before creating the domain so anything could be added as needed by consumers it would probably make fulfilling any custom requirements such as this possible.

**Reply by blecx** [reply]
Thank you very much for posting this hint.  I have a question about that template:

What be need to change n the template is that 
```ruby
 <disk type='file' device='disk'>
```
goes to `disk type='block` 

So may be this can be replaced with (I'm not that experienced in ruby, sorry)
```ruby
<disk type='<%= d[:disk_type] %>'/> device='disk'>
``` 
With disk_type resolves to 'file' or 'block'
This requires a varibale that the template accepts to set this value.

The same needs to be done with the <source ... > tag, but this time we need to change the name of the attribute.

```ruby
 <source file='<%= d[:absolute_path] %>'/>
``` 
my goes to ` <source dev='<%= d[:absolute_path] %>'/>`.  So we may need something like

```ruby
 <source '<%= d[:source_type] %>'/='<%= d[:absolute_path] %>'/>
``` 
with source_type resolves to 'file' of 'dev'
So does this work with the used template machine? 

Second question is, that the file based storage is an image. When using a host device as `/dev/sda` it is not needed to create that image.

Thank you for your help,

Stefan

**Reply by blecx** [reply]
I just digged in the code to find who is filling out the template: 
https://github.com/vagrant-libvirt/vagrant-libvirt/blob/07391048afd9e7a26ebaf633cb4a4baceef63c2c/lib/vagrant-libvirt/config.rb#L619

It took this example 
```ruby
Vagrant.configure("2") do |config|
  config.vm.provider :libvirt do |libvirt|
    libvirt.storage :file, :size => '20G'
    libvirt.storage :file, :size => '40G', :bus => 'scsi', :type => 'raw', :discard => 'unmap', :detect_zeroes => 'on'
  end
end
```
and grepped in the code for  libvirt.storage resolves to (Defined in lib/vagrant-libvirt/config.rb) ` def storage(storage_type, options = {})..`. This is an if - cascade that calles handler, for example for disks `  def _handle_disk_storage(options = {}) ...`
So this seems to create an object that will be pasted into `@disks << disk.

If I got it right with this very little ruby knowlege, it requires a new template and a handler for that. So just adopting the existing template as I tought above is not good enough and will not work.

What do you think?

Thank you for any help, Stefan

**Reply by electrofelix** [reply]
Yes, essentially if you look at the following code, there needs to be additional options added along with sensible defaults that match the existing hardcoded values in the template.

https://github.com/vagrant-libvirt/vagrant-libvirt/blob/07391048afd9e7a26ebaf633cb4a4baceef63c2c/lib/vagrant-libvirt/config.rb#L664-L680

followed by an update to the template at:
https://github.com/vagrant-libvirt/vagrant-libvirt/blob/07391048afd9e7a26ebaf633cb4a4baceef63c2c/lib/vagrant-libvirt/templates/domain.xml.erb#L116-L152

To support setting the new options as needed

**Reply by electrofelix** [reply]
Unfortunately it seems we made a mistake in using `type` for the driver type rather than using `driver_type` and keeping `type` for the disk type. But otherwise, yes your suggested changes for the disk XML are along the right lines, it's just the addition of the options for the storage handler that are needed as well as possible alternative defaults for handling cdroms as well.

Bit rusty on host devices and the corresponding XML, but if you are using one directly, it might be that either there is an easy way to determine if the path is a device or not, in which case it's possible to have some options adjusted without needing the Vagrantfile to provide additional information. Potentially there may be issues around permissions with such devices.

**Reply by blecx** [reply]
First,

thank you very much for this imediate reply. So it seems that it needs some work on the code. But for doing so, I need to learn some more ruby, - as well as learning how to build and test vagrant-libvirt. So it will be a "long run".

So getting better with ruby is not that big deal. Any advice how to build and test vagrant-libvirt? 

Thank you very much,

Stefan
PS. In the mean time i can help with triggers, but this really an ugly work around and only works on a single node Vagrant file, since triggers are all global. And it only works with vagrand up/vagrant destroy, - no halt etc.
If it's of interest I can paste here an example

**Reply by electrofelix** [reply]
The general development steps are at https://github.com/vagrant-libvirt/vagrant-libvirt#development, I personally use `rvm` to manage different ruby versions and also to be able to install gems in an isolated manner so that I can easily clean up afterwards by simply calling `rvm gemset delete <name>`.

Documentation for triggers appears to suggest that it covers halt on a per machine type trigger
* https://www.vagrantup.com/docs/triggers/usage
